### PR TITLE
unix: stop overwritting delayed_error by reporting it duly

### PR DESCRIPTION
**Pull request by deleisha**

### CODE REVIEW

File: src/unix/tcp.c [hunk 1]
```diff
@@ -156,7 +156,9 @@ int uv__tcp_connect(uv_connect_t* req,
   if (err)
     return err;
 
-  handle->delayed_error = 0;
+  if (handle->delayed_error) {
+    return handle->delayed_error;
+  }
```
Comment (bnoordhuis) [inline src/unix/tcp.c]: I would move this check up before the call to `maybe_new_socket()`.  Also, can you drop the braces?

**Comment by bnoordhuis** [comment]
I think this addresses the bind-before-connect case where the bind fails?  Can you add a regression test for that?  Thanks.

File: src/unix/tcp.c (hunk 1 above)
Comment (deleisha) [inline src/unix/tcp.c]: Thanks Ben, Yea, I should have done that

**Comment by deleisha** [comment]
@bnoordhuis 
As Linux always succeeds `bind` when SO_REUSEADDR is set and EADDRINUSE is delayed to `listen`. It is being observed that `connect` does not return this error. 
But tend to believe that BSDs will report EADDRINUSE at bind time. In such scenario, we can think of following options 1) do we write test for those platforms using guard? 
2) using SO_REUSEPORT|SO_REUSEADDR for platforms like MacOS, BUT not for Linux, so as to have similar behav

**Comment by bnoordhuis** [comment]
@deleisha Not sure I follow.  The EADDRINUSE should be deferred to the uv_listen/uv_connect call regardless of whether it was the bind system call or the listen/connect system call that failed.

**Comment by deleisha** [comment]
@bnoordhuis correct, what I was trying to say was that on my workstation, which is Linux, bind does not fail as we are setting SO_REUSEADDR. But I tend to believe it might fail on other system like Mac. 

so we have in code. For linux, gdb shows that `errno is 0` after  `bind` call.
```C 
errno = 0;
  if (bind(tcp->io_watcher.fd, addr, addrlen) && errno != EADDRINUSE) {
    if (errno == EAFNOSUPPORT)
      /* OSX, other BSDs and SunoS fail with EAFNOSUPPORT when binding a
       * socket created with AF_INET to an AF_INET6 address or vice versa. */
      return -EINVAL;
    return -errno;
  }
  tcp->delayed_error = -errno;
```
Will it be okay if we test something like
```C
/* calling uv_connect after bind */
r = uv_connect(..);
/* r == 0 on linux */
ASSERT( r == 0 || r == UV_EADDRINUSE);
```

**Comment by bnoordhuis** [comment]
> what I was trying to say was that on my workstation, which is Linux, bind does not fail as we are setting SO_REUSEADDR. But I tend to believe it might fail on other system like Mac.

Yes, there are platform differences.  Some platforms return EADDRINUSE from bind(), others from listen() or connect().

For server sockets, we return the error from `uv_listen()`.  For client sockets, I'd do it from `uv_tcp_connect()`.  It would be possible to defer the error to the connect callback, like we do for ECONNREFUSED, but I'd opt for simplicity here.

The logic in the test should look like this:
```c
ASSERT(0 == uv_bind(&tcp_handle, ...));
ASSERT(UV_EADDRINUSE == uv_tcp_connect(&tcp_handle, ...));
```
I.e. `uv_tcp_connect()` should unconditionally fail with UV_EADDRINUSE.

**Comment by deleisha** [comment]
Yes, that was my initial thought too. But it does not happen that way. Our thinking will be true when 
`bind` fails with `EADDRINUSE` and `uv_tcp_bind` defers it. 

On linux, bind succeeds returning zero and hence `EADDRINUSE` is not differred bu `uv_tcp_bind`. Hence following code piece fails on linux
> ASSERT(0 == uv_bind(&tcp_handle, ...));
> ASSERT(UV_EADDRINUSE == uv_tcp_connect(&tcp_handle, ...));

What I have learnt from google is that BSDs behave like similar to linux when both `SO_REUSEADDR |
SO_REUSEAPORT` are set. I am thinking we still are having small difference among platforms. Yet I dont have other machine to test.

**Comment by bnoordhuis** [comment]
I think I get what you're saying.  The connect() system call fails with EADDRNOTAVAIL (not EADDRINUSE) on Linux, OS X and FreeBSD and I'm reasonably sure it behaves that way on AIX and Solaris too.

I'd write a test that checks for that error code.  On Windows, `uv_tcp_try_connect` probably has to remap `handle->delayed_error` from WSAEADDRINUSE to WSAEADDRNOTAVAIL in order to pass.

@saghul @piscisaureus Thoughts?

**Comment by deleisha** [comment]
Added the test for connect after bind scenario.

**Comment by deleisha** [comment]
@saghul  @piscisaureus WDYT

**Comment by cjihrig** [comment]
This has been inactive for nearly two years, and needs to be rebased. I'm going to close for now.

**Comment by deleisha** [comment]
Has the issue fixed by later commit? If not, the issue might still be there.

**Comment by saghul** [comment]
@deleisha would you be so kind to double-check and reopen a PR? loks like we dropped the ball here, sorry.

**Pull request closed without merging**